modules/shortcodes/instagram.php

```diff
@@ -177,47 +220,50 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
 	 * @param array  $atts       Instagram Shortcode attributes.
 	 * @param string $passed_url Instagram API URL.
 	 */
-	$response_body_use_cache = apply_filters( 'instagram_cache_oembed_api_response_body', false, $matches, $atts, $passed_url );
-	$response_body           = false;
-	if ( $response_body_use_cache ) {
-		$cache_key     = 'oembed_response_body_' . md5( $url );
-		$response_body = wp_cache_get( $cache_key, 'instagram_embeds' );
-	}
+	return apply_filters( 'instagram_cache_oembed_api_response_body', false, $matches, $atts, $passed_url );
+}
 
-	if ( ! $response_body ) {
-		// Not using cache (default case) or cache miss.
-		$instagram_response = wp_remote_get( $url, array( 'redirection' => 0 ) );
-		if (
-			is_wp_error( $instagram_response )
-			|| 200 !== $instagram_response['response']['code']
-			|| empty( $instagram_response['body'] ) ) {
-			return '<!-- instagram error: invalid instagram resource -->';
+/**
+ * Handles the logic for actually fetching information for an Instagram embed, conditionally proxying the
+ * request through the WordPress.com API if this is not a WordPress.com site.
+ *
+ * @param array $args An array of arguments to pass to the embed API.
+ *
+ * @return mixed A HTML comment (string) if errored or an object.
+ */
+function jetpack_instagram_fetch_embed( $args ) {
+	if ( Constants::is_defined( 'IS_WPCOM' ) && Constants::get_constant( 'IS_WPCOM' ) ) {
+		$url      = esc_url_raw(
+			add_query_arg(
+				$args,
+				'https://api.instagram.com/oembed/'
+			)
+		);
+		$response = wp_remote_get( $url, array( 'redirection' => 0 ) );
+	} else {
+		if ( ! Jetpack::is_active() ) {
+			return '<!-- instagram error: Jetpack is not connected -->';
 		}
 
-		$response_body = json_decode( $instagram_response['body'] );
-		if ( $response_body_use_cache ) {
-			// if caching it is short-lived since this is a "Cache-Control: no-cache" resource.
-			wp_cache_set(
-				$cache_key,
-				$response_body,
-				'instagram_embeds',
-				HOUR_IN_SECONDS + wp_rand( 0, HOUR_IN_SECONDS )
-			);
-		}
+		$response = Client::wpcom_json_api_request_as_blog(
+			'/oembed-proxy/instagram?' . http_build_query( $args ),
+			'2',
+			array(),
+			null,
+			'wpcom'
+		);
 	}
 
-	if ( ! empty( $response_body->html ) ) {
-		wp_enqueue_script(
-			'jetpack-instagram-embed',
-			Assets::get_file_url_for_environment( '_inc/build/shortcodes/js/instagram.min.js', 'modules/shortcodes/js/instagram.js' ),
-			array( 'jquery' ),
-			JETPACK__VERSION,
-			true
-		);
-		return $response_body->html;
+	$response_body = json_decode( wp_remote_retrieve_body( $response ) );
+	if (
+		is_wp_error( $response )
+		|| 200 !== wp_remote_retrieve_response_code( $response )
+		|| empty( $response_body )
+	) {
+		return '<!-- instagram error: invalid instagram resource -->';
```

Comment:
As a note, as part of the refactor, this error message is effectively overridden currently in the calling function currently. That's not a huge deal, but should probably be addressed.